src/inmanta/server/services/orchestrationservice.py

```diff
@@ -468,117 +395,155 @@ async def delete_version(self, env: data.Environment, version_id: int) -> Apiret
             if "requires" not in attributes:
-                LOGGER.warning("Received resource without requires attribute (%s)" % res_obj.resource_id)
+                LOGGER.warning("Received resource without requires attribute (%s)", res_obj.resource_id)
             else:
                 # Collect all requires as resource_ids instead of resource version ids
                 cleaned_requires = []
                 for req in attributes["requires"]:
                     rid = Id.parse_id(req)
-                    provides_tree[rid.resource_str()].append(resc_id)
+                    all_requires.add(rid.resource_str())
                     if rid.get_agent_name() != agent:
                         # it is a CAD
                         cross_agent_dep.append((res_obj, rid))
                     cleaned_requires.append(rid.resource_str())
                 attributes["requires"] = cleaned_requires
-        resource_ids = {res.resource_id for res in resource_objects}
-        superfluous_ids = set(resource_sets.keys()) - resource_ids
+
+            rid_to_resource[res_obj.resource_id] = res_obj
+
+        # hook up all CADs
+        for f, t in cross_agent_dep:
+            res_obj = rid_to_resource[t.resource_str()]
+            res_obj.provides.append(f.resource_id)
+
+        rids = set(rid_to_resource.keys())
+
+        # Sanity checks
+        superfluous_ids = set(resource_sets.keys()) - rids
         if superfluous_ids:
             raise BadRequest(
                 "The following resource ids provided in the resource_sets parameter are not present "
                 f"in the resources list: {', '.join(superfluous_ids)}"
             )
-        requires_ids = set(provides_tree.keys())
-        if not requires_ids.issubset(resource_ids):
+        if not all_requires.issubset(rids):
             raise BadRequest(
                 "The model should have a dependency graph that is closed and no dangling dependencies:"
-                f" {requires_ids-resource_ids}"
+                f" {all_requires - rids}"
             )
-        # hook up all CADs
-        for f, t in cross_agent_dep:
-            res_obj = rv_dict[t.resource_str()]
-            res_obj.provides.append(f.resource_id)
+
+        return rid_to_resource
+
+    def _get_skipped_for_undeployable(
+        self, resources: abc.Sequence[data.Resource], undeployable_ids: abc.Sequence[ResourceIdStr]
+    ) -> abc.Sequence[ResourceIdStr]:
+        """
+        Return the resources that are skipped_for_undeployable given the full set of resources and
+        the resource ids of the resources that are undeployable.
+
+        :param resources: All resources in the model.
+        :param undeployable_ids: The ids of the resource that are undeployable.
+        """
+        # Build up provides tree
+        provides_tree: Dict[ResourceIdStr, List[ResourceIdStr]] = defaultdict(lambda: [])
+        for r in resources:
+            if "requires" in r.attributes:
```

Comment:
All methods only consider `"requires"`. Don't our DAO's attributes' have a `"provides"` attribute? Do we convert those into requires at some point?

Reply:
As discussed, resources sent to the server only contain the requires relationship and not the provides relationship. The compiler turns the provides into requires.